components/Canvas.tsx

```diff
@@ -16,43 +16,47 @@ export function useCanvas() {
 }
 
 export function useContext2d() {
-  const context2d = useContext(Context2DContext);
+  const context2d = useContext(Context2dContext);
   if (!context2d) {
     throw new Error("useContext must be used within a Canvas component");
   }
 
   return context2d;
 }
 
-export function Canvas({ children }: { children: preact.ComponentChildren }) {
-  const canvasRef = useRef<HTMLCanvasElement>(null);
-  const contextRef = useRef<CanvasRenderingContext2D | null>(null);
-  const [isMounted, setIsMounted] = useState(false);
+export type CanvasProps = { children?: preact.JSX.Element | null };
+
+export function Canvas({ children }: CanvasProps) {
+  const [canvasElement, setCanvasElement] = useState<HTMLCanvasElement | null>(
+    null
+  );
   const [width, setWidth] = useState(0);
   const [height, setHeight] = useState(0);
-  useLayoutEffect(() => {
+
+  useEffect(() => {
     const handleResize = () => {
       setWidth(innerWidth);
       setHeight(innerHeight);
     };
-    contextRef.current = canvasRef.current?.getContext("2d") ?? null;
-    setIsMounted(true);
     addEventListener("resize", handleResize);
     handleResize();
     return () => {
-      setIsMounted(false);
       removeEventListener("resize", handleResize);
     };
-  }, []);
+  }, [canvasElement]);
+  const context2d = useMemo(
```

Comment:
@RoryDuncan, do you know if the getContext("2d") is something that needs to be cached? In my testing, it seemed to work even when it wasn't, but doing it here just in case I don't know what I'm doing.

Reply:
Pretty sure repeated calls return the same instance. Probably doesn't matter, but cache'ing won't hurt nothin'.